docs/developers/local-rules.md

```diff
@@ -0,0 +1,212 @@
+set up correctly](../users/installation.md#sbt).
+
+Although it is possible to define your rules in a Scala binary version that
+does not match your build, it is highly recommended that you align them via:
+
+```diff
+ // build.sbt
++scalafixScalaBinaryVersion in ThisBuild :=
++  CrossVersion.binaryScalaVersion(scalaVersion.value)
+```
+
+> Note that any potential external Scalafix rule, loaded with the
+> `scalafixDependencies` setting key, must be built and published against the
+> same Scala binary version.
+
+## Single-project sbt build
+
+If you do not have any sub-project in your build, custom rules should be
+defined under the Scalafix Ivy configuration sources directory, which by
+default is `src/scalafix/scala`.
+
+```diff
+ repository
+ ├── build.sbt
+ ├── project
+ │   └── plugins.sbt
+ └── src
+     ├── main
+     │   └── ...
+     ├── test
+     │   └── ...
++    └── scalafix
++       ├── resources/META-INF/services
++       │   └── scalafix.v1.Rule // not mandatory, but allows syntax
++       |                        // `MyRule1` over `class:fix.MyRule1`
++       └── scala/fix
++           ├── MyRule1.scala
++           ├── MyRule2.scala
++           └── ...
+```
+
+```diff
+ // build.sbt
++libraryDependencies +=
++  "ch.epfl.scala" %%
++    "scalafix-core" %
++    _root_.scalafix.sbt.BuildInfo.scalafixVersion %
++    ScalafixConfig
+```
+
+```bash
+$ sbt
+> scalafix MyRule1 MyRule2
+```
+
+## Multi-project sbt build
+
+If your build contains several sub-projects, rules can be defined either:
+
+1. within [one of them](#within-a-given-sub-project), which is simpler but
+   make them impossible to test and complicate sharing rules across
+   sub-projects;
+1. under [their own sub-project(s)](#as-a-separate-sub-project), allowing
+   usage of [scalafix-testkit](setup.md).
+
+### Within a given sub-project
+
+In a large build with different code owners across sub-projects, it may
+be useful to define rules within a given sub-project, without impacting the
+global build definition. This helps to iterate on them before potentially
+promoting them to the [build-level](#as-a-separate-sub-project).
+
+```diff
+ repository
+ ├── build.sbt
+ ├── project
+ │   └── plugins.sbt
+ └── service1
+ │   └── src
+ │       ├── main
+ │       │   └── ...
+ │       ├── test
+ │       │   └── ...
++│       └── scalafix
++│           ├── resources/META-INF/services
++│           │   └── scalafix.v1.Rule // not mandatory, but allows syntax
++|           |                        // `MyRule1` over `class:fix.MyRule1`
++│           └── scala/fix
++│               ├── MyRule1.scala
++│               ├── MyRule2.scala
++│               └── ...
+ ├── service2
+ │   └── ...
+ └── service3
+     └── ...
+```
+
+```diff
+ // build.sbt
+ lazy val service1 = project
++  .settings(
++    libraryDependencies +=
++      "ch.epfl.scala" %%
++        "scalafix-core" %
++        _root_.scalafix.sbt.BuildInfo.scalafixVersion %
++        ScalafixConfig
```

Comment:
I guess we could make this automatic, as long as we update https://github.com/scalacenter/sbt-scalafix/blob/60199fdae7bb9c56b9d09969b0889923a6b39108/src/main/scala/scalafix/sbt/ScalafixPlugin.scala#L201-L202 to avoid always invalidating the default tool classpath.

Reply:
I don't get it. It's normal to add scalafix-core to library dependencies?

Reply:
It would be OK to add scalafix-core to library dependencies *in the ScalafixConfig* configuration, since anything there is meant to be rules, which require types from there.

Another downside is that it would potentially expose classpath collisions in IntelliJ mentioned in https://github.com/scalacenter/scalafix/pull/1173/files#r443266186 to all plugin users, whether or not they use local rules.